File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0, ans = 0;
        set<char> window;

        for (int r = 0 ; r < s.size() ; r++){
            while (l < r && window.find(s[r]) != window.end()){
                window.erase(s[l]);
                l++;
            }

            window.insert(s[r]);
            ans = max(ans, (int)window.size());
            
        }

        return ans;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        ans = r

        while l <= r:
            rate = (l+r) // 2

            time_taken = 0
            for p in piles:
                time_taken += (p+rate-1) // rate

            if time_taken <= h:
                ans = rate
                r = rate-1
            else:
                l = rate+1
        return ans
            

